thewalrus/quantum.py

```diff
@@ -572,6 +573,72 @@ def state_vector(
     return psi
 
 
+def mean_number_of_clicks(A):
+    r"""
+    Given an adjacency matrix it calculates the mean number of clicks.
+    For this to make sense the user must provide a matrix with singular values
+    less than or equal to one. See Appendix A.3 of arxiv.org/abs/1902.00462
+    by Banchi et al.
+
+    Args:
+        A (array): rescaled adjacency matrix
+    Returns:
+        float: mean number of clicks
+    """
+    n, _ = A.shape
+    idn = np.identity(n)
+    X = np.block([[0 * idn, idn], [idn, 0 * idn]])
+    B = np.block([[A, 0 * A], [0 * A, np.conj(A)]])
+    Q = np.linalg.inv(np.identity(2 * n) - X @ B)
+    meanc = 1.0 * n
+    for i in range(n):
+        det_val = np.real(Q[i, i] * Q[i + n, i + n] - Q[i + n, i] * Q[i, i + n])
+        meanc -= 1.0 / np.sqrt(det_val)
+    return meanc
+
+
+def find_scaling_adjacency_matrix_torontonian(A, c_mean):
+    r""" Returns the scaling parameter by which the adjacency matrix A
+    should be rescaled so that the Gaussian state that endodes it has
+    give a mean number of clicks equal c_mean when measured with
+    threshold detectors.
+
+    Args:
+        A (array): adjacency matrix
+        c_mean (float): mean photon number of the Gaussian state
+
+    Returns:
+        float: scaling parameter
+    """
+    n, _ = A.shape
+    if 0 >= c_mean > n:

```

Comment:
Does this make sense? The first half implies c_mean is negative, but `n` is a positive number, so this should never be the case

Reply:
Nop it does not